rig/machine_control/scp_protocol.py

```diff
@@ -0,0 +1,411 @@
+            timer_handle=timer_handle,
+            n_tries=n_tries,
+        )
+
+    def on_timeout(self, seq_num):
+        """Callback when an SCP packet response times out.
+
+        Retransmit or drop the packet.
+        """
+        out_req = self.outstanding.pop(seq_num)
+        new_n_tries = out_req.n_tries + 1
+
+        if out_req.queued_request.future.cancelled():
+            # Task cancelled, no point in re-transmitting. Handle some new
+            # packets instead
+            self._process_queue()
+        elif new_n_tries > self.n_tries:
+            # The packet has been retransmitted too many times, drop it
+            out_req.queued_request.future.set_exception(TimeoutError(
+                "Exceeded {} attempts trying to transmit packet.".format(
+                    self.n_tries)))
+
+            # Since a previously outstanding packet has now been serviced, a
+            # new packet from the queue can potentially be processed.
+            self._process_queue()
+        else:
+            # Retransmit
+            self._send_packet(out_req.queued_request, new_n_tries)
+
+    def datagram_received(self, data, addr):
+        """Callback on a datagram arriving from the transport.
+
+        Process the incoming SCP response.
+        """
+        # XXX: Silently ignores packets whose sequence number is not
+        # recognised.
+        seq_num = SCPPacket.from_bytestring(data[2:]).seq
+        if seq_num in self.outstanding:
+            out_req = self.outstanding.pop(seq_num)
+
+            # Cancel the timeout
+            out_req.timer_handle.cancel()
+
+            # Fully-unpack the packet
+            packet = SCPPacket.from_bytestring(
+                data[2:], n_args=out_req.queued_request.expected_args)
+
+            # Handle the response
+            if out_req.queued_request.future.cancelled():
+                # If the future was cancelled, there's no need to do anything
+                # with the response: simply ignore it.
+                pass
+            if packet.cmd_rc not in self._error_codes:
+                # The response is good, send it to the original requester
+                if not out_req.queued_request.future.cancelled():
+                    out_req.queued_request.future.set_result(packet)
+            else:
+                # The CMD_RC indicates that the command failed. Raise an
+                # appropriate exception detailing the original packet sent.
+                orig_packet = SCPPacket.from_bytestring(
+                    out_req.queued_request.data[2:])
+                if not out_req.queued_request.future.cancelled():
+                    out_req.queued_request.future.set_exception(
+                        self._error_codes[packet.cmd_rc](
+                            "Error in response to packet with arguments: "
+                            "cmd_rc={}, arg1={}, arg2={}, arg3={}; "
+                            "sent to core ({},{},{}).".format(
```

Comment:
For reference, having implemented this I don't like it. I'd rather pass the orig_packet and the received failure and let the exception decide how to best display the error...